addons/website_sale/controllers/main.py

```diff
@@ -1220,14 +1220,16 @@ def _get_products_recently_viewed(self):
         """
         max_number_of_product_for_carousel = 12
         visitor = request.env['website.visitor']._get_visitor_from_request()
+        company = request.website.company_id or request.env.company
         if visitor:
             excluded_products = request.website.sale_get_order().mapped('order_line.product_id.id')
+            allowed_products = self.env["product.product"].search([('website_published', '=', True), '|', ('company_id', '=', company.id), ('company_id', '=', False), ('id', 'not in', excluded_products)])
             products = request.env['website.track'].sudo().read_group(
-                [('visitor_id', '=', visitor.id), ('product_id', '!=', False), ('product_id.website_published', '=', True), ('product_id', 'not in', excluded_products)],
-                ['product_id', 'visit_datetime:max'], ['product_id'], limit=max_number_of_product_for_carousel, orderby='visit_datetime DESC')
+                [('visitor_id', '=', visitor.id), ('product_id', '!=', False)],
+                ['product_id', 'visit_datetime:max'], ['product_id'], orderby='visit_datetime DESC')
             products_ids = [product['product_id'][0] for product in products]
             if products_ids:
-                viewed_products = request.env['product.product'].with_context(display_default_code=False).browse(products_ids)
+                viewed_products = (request.env['product.product'].with_context(display_default_code=False).browse(products_ids) & allowed_products)[:max_number_of_product_for_carousel]
```

Comment:
Well the idea is to avoid prefetching doing two really simple query and then just doing a recordset operation. So in the DB side we'd only get two queries.

With the first one we get a recordset. It can be huge, but we won't be prefetching anything. We could even do a `_search` and then a browse to avoid top level overrides:

```python
allowed_products = request.env["product.product"]._search([('website_published', '=', True), '|', ('company_id', '=', company.id), ('company_id', '=', False), ('id', 'not in', excluded_products)])
```

The second query will be very simple as well. This only gets `ids` as well and it just gets another list of `ids`. No prefetching nor anything:

```python
            products = request.env['website.track'].sudo().read_group(
                [('visitor_id', '=', visitor.id), ('product_id', '!=', False)],
                ['product_id', 'visit_datetime:max'], ['product_id'], orderby='visit_datetime DESC')
            products_ids = [product['product_id'][0] for product in products]
```

I just put this query in first place indeed, as we could discard the first query if we don't get any result here.

And then, for the final trick, we just make an `&` recordset operation. No more querys involved, no prefetching. And simply get the `max_number_of_product_for_carousel` first records.

```python
viewed_products = (request.env['product.product'].with_context(display_default_code=False).browse(products_ids) & request.env['product.product'].browse(allowed_products))[:max_number_of_product_for_carousel]
```

`max_number_of_product_for_carousel` by the way is a hardcoded value: 
https://github.com/oca/ocb/blob/65900657bcbc19254aaf0d5b038692686f1dbc3f/addons/website_sale/controllers/main.py#L1221

Reply:
Some performance tests following my approach. Certainly in very large catalogs (hundreds of thousands of products) it would perform poorly while in regular catalogs (thousands of products) it'd be acceptable.


```python
import random
import timeit
from odoo.tools import OrderedSet


def test_performance(allowed=100, viewed=100, limit_records=12):
    """Simulate orm __and__ operation over two recordsets""""
    allowed_product_ids = random.sample(range(1, 1000000), allowed)
    viewed_products = random.sample(range(1, 1000000), viewed)
    list(OrderedSet(id for id in viewed_products if id in allowed_product_ids))[:limit_records]

# From worst performance to best. The biggest the catalog and the biggest the intersection, the worst performs.
timeit.timeit(lambda: test_performance(900000, 1000), number=1) # 28.113155835017096
timeit.timeit(lambda: test_performance(90000, 1000), number=1) # 4.8130702060007025
timeit.timeit(lambda: test_performance(900000, 100), number=1) # 3.215424895985052
timeit.timeit(lambda: test_performance(9000, 9000), number=1) # 0.8743320660141762
timeit.timeit(lambda: test_performance(90000, 100), number=1) # 0.60079662600765
timeit.timeit(lambda: test_performance(9000, 1000), number=1) # 0.11547618900658563
timeit.timeit(lambda: test_performance(1000, 1000), number=1) # 0.023072323005180806
timeit.timeit(lambda: test_performance(1000, 100), number=1) # 0.007160857989219949
```

Reply:
Changed. Take a look

Reply:
Hello 👋 
I wrote a patch against your branch here, with an added test, I prefer keeping the diff inside the controller as minimal as possible: https://gist.github.com/Williambraecky/ec4c155e622aa6630ae677d029a6eeb7
Let me know if this is good for you.
Cheers

Reply:
I can't apply the patch. Can you PR to my branch?

Reply:
Hello @chienandalu, done over at https://github.com/Tecnativa/odoo/pull/12/files